File: tester/src/test/python/spinal/OperatorTester/OperatorTester.py
import cocotb
from cocotb.result import TestFailure
from cocotb.triggers import Timer

from cocotblib.misc import randSignal, truncUInt, sint, truncSInt
import random

def check(signal,bitCount,value):
    if len(signal) != bitCount or int(signal) != truncUInt(value, bitCount):
        raise TestFailure("FAIL %s    %d != %d" % (signal._path, int(signal), truncUInt(value, bitCount)))

def checkSigned(signal,bitCount,value):
    if len(signal) != bitCount or sint(signal) != truncSInt(value,bitCount):
        raise TestFailure("FAIL %s    %d != %d" % (signal._path, sint(signal), truncSInt(value,bitCount)))

def notZero(x):
    if x == 0:
        return 1
    else:
        return x

def randSignalNotZero(that):
    value = random.getrandbits(len(that))
    if(value == 0):
        value += 1
    that <= value

@cocotb.test()
def test1(dut):
    dut.log.info("Cocotb test boot")

    # itemDir = dut.uint4.__dict__
    #
    # for i in itemDir:
    #     print '{0}  :  {1}'.format(i, itemDir[i])

    for i in range(0,2000):
        randSignal(dut.uint4)
        randSignal(dut.uint8)
        randSignalNotZero(dut.uint4NotZero)
        randSignalNotZero(dut.uint8NotZero)
        randSignal(dut.uint32)
        randSignal(dut.sint4)
        randSignal(dut.sint8)
        randSignalNotZero(dut.sint4NotZero)
        randSignalNotZero(dut.sint8NotZero)
        randSignal(dut.sint32)
        randSignal(dut.bits4)
        randSignal(dut.bits8)
        randSignal(dut.bits32)
        randSignal(dut.boolA)
        randSignal(dut.boolB)
        randSignal(dut.boolC)
        yield Timer(1000)
        uint4 = int(dut.uint4)
        uint8 = int(dut.uint8)
        uint32 = int(dut.uint32)
        sint4 = sint(dut.sint4)
        sint8 = sint(dut.sint8)
        sint32 = sint(dut.sint32)
        bits4 = int(dut.bits4)
        bits8 = int(dut.bits8)
        bits32 = int(dut.bits32)
        boolA = int(dut.boolA)
        boolB = int(dut.boolB)
        boolC = int(dut.boolC)
        uint4NotZero =  int(dut.uint4NotZero)
        uint8NotZero =  int(dut.uint8NotZero)
        sint4NotZero = sint(dut.sint4NotZero)
        sint8NotZero = sint(dut.sint8NotZero)



        check(dut.boolMux,1, boolA if boolC else boolB)
        check(dut.bitsBsMux,8, bits8 if boolC else bits4)
        check(dut.uintBsMux,8, uint8 if boolC else uint4)
        check(dut.sintBsMux,8, sint8 if boolC else sint4)
        check(dut.bitsSbMux,8, bits4 if boolC else bits8)
        check(dut.uintSbMux,8, uint4 if boolC else uint8)
        check(dut.sintSbMux,8, sint4 if boolC else sint8)
        check(dut.stateNativeMux            ,2, 0 if boolC else 1)
        check(dut.stateBinarySequancialMux  ,2, 0 if boolC else 1)
        check(dut.stateBinaryOneHotMux      ,3, 1 if boolC else 2)

        check(dut.uintNot, 4, ~uint4)
        check(dut.uintShiftLeftInt  , 12, uint8 << 4)
        check(dut.uintShiftLeftUint , 23, uint8 << uint4)
        check(dut.uintShiftRightInt , 4, uint8 >> 4)
        check(dut.uintShiftRightUint, 8, uint8 >> uint4)
        check(dut.uintShiftLeftIntFixedWidth  , 8, uint8 << 4)
        check(dut.uintShiftLeftUintFixedWidth , 8, uint8 << uint4)
        check(dut.uintShiftRightIntFixedWidth , 8, uint8 >> 4)
        check(dut.uintShiftRightUintFixedWidth, 8, uint8 >> uint4)

        check(dut.sintNot, 4, ~sint4)
        checkSigned(dut.sintMinus, 4, -sint4)
        check(dut.sintShiftLeftInt  , 12, sint8 << 4)
        check(dut.sintShiftLeftUint , 23, sint8 << uint4)
        check(dut.sintShiftRightInt , 4, sint8 >> 4)
        checkSigned(dut.sintShiftRightUint, 8, sint8 >> uint4)
        check(dut.sintShiftLeftIntFixedWidth  , 8, sint8 << 4)
        check(dut.sintShiftLeftUintFixedWidth , 8, sint8 << uint4)
        checkSigned(dut.sintShiftRightIntFixedWidth , 8, sint8 >> 4)
        checkSigned(dut.sintShiftRightUintFixedWidth, 8, sint8 >> uint4)

        check(dut.bitsNot, 4, ~bits4)
        check(dut.bitsShiftLeftInt  , 12, bits8 << 4)
        check(dut.bitsShiftLeftUint , 23, bits8 << uint4)
        check(dut.bitsShiftRightInt , 4, bits8 >> 4)
        check(dut.bitsShiftRightUint, 8, bits8 >> uint4)
        check(dut.bitsShiftLeftIntFixedWidth  , 8, bits8 << 4)
        check(dut.bitsShiftLeftUintFixedWidth , 8, bits8 << uint4)
        check(dut.bitsShiftRightIntFixedWidth , 8, bits8 >> 4)
        check(dut.bitsShiftRightUintFixedWidth, 8, bits8 >> uint4)



        check(dut.uintSbEquals,         1, uint4 == uint8)
        check(dut.uintSbNotEquals,      1, uint4 != uint8)

        check(dut.uintSbAdd, 8, uint4 + uint8)
        check(dut.uintSbSub, 8, uint4 - uint8)
        check(dut.uintSbMul, 12, uint4 * uint8)
        if uint8NotZero != 0:
            check(dut.uintSbDiv, 4, uint4 // uint8NotZero)
            check(dut.uintSbRem, 4, uint4 % uint8NotZero)
        check(dut.uintSbAnd, 8, uint4 & uint8)
        check(dut.uintSbOr , 8, uint4 | uint8)
        check(dut.uintSbXor, 8, uint4 ^ uint8)

        check(dut.uintSbSmaller      , 1, uint4 < uint8)
        check(dut.uintSbSmallerEquals, 1, uint4 <= uint8)
        check(dut.uintSbBigger       , 1, uint4 > uint8)
        check(dut.uintSbBiggerEquals , 1, uint4 >= uint8)




        check(dut.sintSbEquals, 1, sint4 == sint8)
        check(dut.sintSbNotEquals, 1, sint4 != sint8)

        checkSigned(dut.sintSbAdd, 8, sint4 + sint8)
        checkSigned(dut.sintSbSub, 8, sint4 - sint8)
        checkSigned(dut.sintSbMul, 12, sint4 * sint8)
        if sint8NotZero != 0 and sint4 > 0 and sint8NotZero > 0:
            checkSigned(dut.sintSbDiv, 4,  int(sint4//sint8NotZero))
            checkSigned(dut.sintSbRem, 4,  int(sint4% sint8NotZero))

        checkSigned(dut.sintSbAnd, 8, sint4 & sint8)
        checkSigned(dut.sintSbOr, 8, sint4 | sint8)
        checkSigned(dut.sintSbXor, 8, sint4 ^ sint8)

        check(dut.sintSbSmaller, 1, sint4 < sint8)
        check(dut.sintSbSmallerEquals, 1, sint4 <= sint8)
        check(dut.sintSbBigger, 1, sint4 > sint8)
        check(dut.sintSbBiggerEquals, 1, sint4 >= sint8)




        check(dut.bitsSbEquals, 1, bits4 == bits8)
        check(dut.bitsSbNotEquals, 1, bits4 != bits8)

        check(dut.bitsSbAnd, 8, bits4 & bits8)
        check(dut.bitsSbOr, 8, bits4 | bits8)
        check(dut.bitsSbXor, 8, bits4 ^ bits8)








        check(dut.uintBsEquals,         1, uint8 == uint4)
        check(dut.uintBsNotEquals,      1, uint8 != uint4)

        check(dut.uintBsAdd, 8, uint8 + uint4)
        check(dut.uintBsSub, 8, uint8 - uint4)
        check(dut.uintBsMul, 12, uint8 * uint4)
        if uint4NotZero != 0:
            check(dut.uintBsDiv, 8, uint8 // uint4NotZero)
            check(dut.uintBsRem, 8, uint8 % uint4NotZero)
        check(dut.uintBsAnd, 8, uint8 & uint4)
        check(dut.uintBsOr , 8, uint8 | uint4)
        check(dut.uintBsXor, 8, uint8 ^ uint4)

        check(dut.uintBsSmaller      , 1, uint8 < uint4)
        check(dut.uintBsSmallerEquals, 1, uint8 <= uint4)
        check(dut.uintBsBigger       , 1, uint8 > uint4)
        check(dut.uintBsBiggerEquals , 1, uint8 >= uint4)




        check(dut.sintBsEquals, 1, sint8 == sint4)
        check(dut.sintBsNotEquals, 1, sint8 != sint4)

        checkSigned(dut.sintBsAdd, 8, sint8 + sint4)
        checkSigned(dut.sintBsSub, 8, sint8 - sint4)
        checkSigned(dut.sintBsMul, 12, sint8 * sint4)
        if sint4NotZero != 0 and sint8 > 0 and sint4NotZero > 0:
            checkSigned(dut.sintBsDiv, 8,  int(sint8// sint4NotZero))
            checkSigned(dut.sintBsRem, 8,  int(sint8%  sint4NotZero))

        checkSigned(dut.sintBsAnd, 8, sint8 & sint4)
        checkSigned(dut.sintBsOr, 8, sint8 | sint4)
        checkSigned(dut.sintBsXor, 8, sint8 ^ sint4)

        check(dut.sintBsSmaller, 1, sint8 < sint4)
        check(dut.sintBsSmallerEquals, 1, sint8 <= sint4)
        check(dut.sintBsBigger, 1, sint8 > sint4)
        check(dut.sintBsBiggerEquals, 1, sint8 >= sint4)




        check(dut.bitsBsEquals, 1, bits8 == bits4)
        check(dut.bitsBsNotEquals, 1, bits8 != bits4)

        check(dut.bitsBsAnd, 8, bits8 & bits4)
        check(dut.bitsBsOr, 8, bits8 | bits4)
        check(dut.bitsBsXor, 8, bits8 ^ bits4)

        check(dut.bitsCat, 12, bits8 * 16 +  bits4)


        check(dut.boolEquals, 1, boolA == boolB)
        check(dut.boolNotEquals, 1, boolA != boolB)

        check(dut.boolAnd, 1, boolA & boolB)
        check(dut.boolOr, 1, boolA | boolB)
        check(dut.boolXor, 1, boolA ^ boolB)

        check(dut.uintAsBits,8,int(dut.uint8))
        check(dut.uintAsSint,8,int(dut.uint8))
        check(dut.sintAsBits,8,int(dut.sint8))
        check(dut.sintAsUint,8,int(dut.sint8))
        check(dut.bitsAsUint,8,int(dut.bits8))
        check(dut.bitsAsSint,8,int(dut.bits8))
        check(dut.boolAsBits,1,int(dut.boolA))
        check(dut.boolAsUInt,1,int(dut.boolA))
        check(dut.boolAsSInt,1,int(dut.boolA))

        check(dut.bitsResizeBigger  ,16,bits8)
        check(dut.bitsResizeSmaller ,4 ,bits8)
        check(dut.uintResizeBigger  ,16,uint8)
        check(dut.uintResizeSmaller ,4 ,uint8)
        check(dut.sintResizeBigger  ,16,sint8)
        check(dut.sintResizeSmaller ,4 ,sint8)

        bits8StateSeq = (bits8 & 3)
        bits8StateHO = (bits8 & 7)

        if bits8StateSeq <= 2:
            check(dut.stateNativeBits,2,bits8)
            check(dut.stateBinarySequancialBits, 2, bits8)
            check(dut.stateNativeIsA,1,bits8StateSeq == 0)
            check(dut.stateNativeIsB, 1, bits8StateSeq == 1)
            check(dut.stateNativeIsC, 1, bits8StateSeq == 2)
            check(dut.stateNativeIsNotA,1,bits8StateSeq != 0)
            check(dut.stateNativeIsNotB, 1, bits8StateSeq != 1)
            check(dut.stateNativeIsNotC, 1, bits8StateSeq != 2)
            check(dut.stateBinarySequancialIsA,1,bits8StateSeq == 0)
            check(dut.stateBinarySequancialIsB, 1, bits8StateSeq == 1)
            check(dut.stateBinarySequancialIsC, 1, bits8StateSeq == 2)
            check(dut.stateBinarySequancialIsNotA,1,bits8StateSeq != 0)
            check(dut.stateBinarySequancialIsNotB, 1, bits8StateSeq != 1)
            check(dut.stateBinarySequancialIsNotC, 1, bits8StateSeq != 2)

        if bits8StateHO == 1 or bits8StateHO == 2 or bits8StateHO == 4:
            check(dut.stateBinaryOneHotBits, 3, bits8)
            check(dut.stateBinaryOneHotIsA,1,bits8StateHO == 1)
            check(dut.stateBinaryOneHotIsB, 1, bits8StateHO == 2)
            check(dut.stateBinaryOneHotIsC, 1, bits8StateHO == 4)
            check(dut.stateBinaryOneHotIsNotA,1,bits8StateHO != 1)
            check(dut.stateBinaryOneHotIsNotB, 1, bits8StateHO != 2)
            check(dut.stateBinaryOneHotIsNotC, 1, bits8StateHO != 4)

        check(dut.bitsAggregateFixed, 8, (3 << 5) | ((bits8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))
        check(dut.uintAggregateFixed, 8, (3 << 5) | ((uint8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))
        check(dut.sintAggregateFixed, 8, (3 << 5) | ((sint8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))


        check(dut.bitsAggregateUnfixedWidthFixedDefault, 8, (3 << 5) | ((bits8 & 3) << 3) |(boolA << 2) | (3 << 1) | (1 << 0))
        check(dut.uintAggregateUnfixedWidthFixedDefault, 8, (3 << 5) | ((uint8 & 3) << 3) |(boolA << 2) | (3 << 1) | (1 << 0))
        check(dut.sintAggregateUnfixedWidthFixedDefault, 8, (3 << 5) | ((sint8 & 3) << 3) |(boolA << 2) | (3 << 1) | (1 << 0))


        check(dut.bitsAggregateUnfixedWidthUnfixedDefault, 8, (3 << 5) | ((bits8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))
        check(dut.uintAggregateUnfixedWidthUnfixedDefault, 8, (3 << 5) | ((uint8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))
        check(dut.sintAggregateUnfixedWidthUnfixedDefault, 8, (3 << 5) | ((sint8 & 3) << 3) |(boolA << 2) | (boolA << 1) | (1 << 0))

        bits27 = bits32 & ((1 << 27) - 1)
        uint5  = uint8 & 0x1F
        uint5mod27 = uint5 % 27
        rotateLeftValue = (bits27 << uint5mod27) | (bits27 >> (27-uint5mod27))
        rotateRightValue = (bits27 >> uint5mod27) | (bits27 << (27-uint5mod27))
        check(dut.bitsRotateLeftUInt,27,rotateLeftValue)
        check(dut.uintRotateLeftUInt, 27, rotateLeftValue)
        check(dut.sintRotateLeftUInt, 27, rotateLeftValue)
        check(dut.bitsRotateRightUInt,27,rotateRightValue)
        check(dut.uintRotateRightUInt, 27, rotateRightValue)
        check(dut.sintRotateRightUInt, 27, rotateRightValue)

    dut.log.info("Cocotb test done")

File: core/src/test/scala/spinal/core/OperatorTester.scala
/*
 * SpinalHDL
 * Copyright (c) Dolu, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

package spinal.core

import spinal.tester.SpinalTesterCocotbBase

object OperatorTester {
  object State extends SpinalEnum{
    val a,b,c = newElement
  }


  class OperatorTester extends Component {
    val uint4 = in UInt(4 bits)
    val uint8 = in UInt(8 bits)
    val uint32 = in UInt(32 bits)
    val sint4 = in SInt(4 bits)
    val sint8 = in SInt(8 bits)
    val sint32 = in SInt(32 bits)
    val bits4 = in Bits(4 bits)
    val bits8 = in Bits(8 bits)
    val bits32 = in Bits(32 bits)
    val boolA,boolB,boolC = in Bool()
    val uint4NotZero = in UInt(4 bits)
    val uint8NotZero = in UInt(8 bits)
    val sint4NotZero = in SInt(4 bits)
    val sint8NotZero = in SInt(8 bits)


//    def notZero(x : UInt) = ((x === 0) ? (x + 1) | x)
//    def notZero(x : SInt) = ((x === 0) ? (x + 1) | x)


    val boolMux = out(Mux(boolC,boolA,boolB))

    val bitsBsMux = out(Mux(boolC,bits8,bits4.resized))
    val uintBsMux = out(Mux(boolC,uint8,uint4))
    val sintBsMux = out(Mux(boolC,sint8,sint4))
    val bitsSbMux = out(Mux(boolC,bits4.resized,bits8))
    val uintSbMux = out(Mux(boolC,uint4,uint8))
    val sintSbMux = out(Mux(boolC,sint4,sint8))


    val uintNot = out(~uint4)
    val uintShiftLeftInt = out(uint8 << 4)
    val uintShiftLeftUint = out(uint8 << uint4)
    val uintShiftRightInt = out(uint8 >> 4)
    val uintShiftRightUint = out(uint8 >> uint4)
    val uintShiftLeftIntFixedWidth   = out(uint8 |<< 4)
    val uintShiftLeftUintFixedWidth  = out(uint8 |<< uint4)
    val uintShiftRightIntFixedWidth  = out(uint8 |>> 4)
    val uintShiftRightUintFixedWidth = out(uint8 |>> uint4)

    val sintNot = out(~sint4)
    val sintShiftLeftInt = out(sint8 << 4)
    val sintShiftLeftUint = out(sint8 << uint4)
    val sintShiftRightInt = out(sint8 >> 4)
    val sintShiftRightUint = out(sint8 >> uint4)
    val sintShiftLeftIntFixedWidth   = out(sint8 |<< 4)
    val sintShiftLeftUintFixedWidth  = out(sint8 |<< uint4)
    val sintShiftRightIntFixedWidth  = out(sint8 |>> 4)
    val sintShiftRightUintFixedWidth = out(sint8 |>> uint4)
    val sintMinus = out(-sint4)

    val bitsNot = out(~bits4)
    val bitsShiftLeftInt = out(bits8 << 4)
    val bitsShiftLeftUint = out(bits8 << uint4)
    val bitsShiftRightInt = out(bits8 >> 4)
    val bitsShiftRightUint = out(bits8 >> uint4)
    val bitsShiftLeftIntFixedWidth = out(bits8 |<< 4)
    val bitsShiftLeftUintFixedWidth = out(bits8 |<< uint4)
    val bitsShiftRightIntFixedWidth = out(bits8 |>> 4)
    val bitsShiftRightUintFixedWidth = out(bits8 |>> uint4)

    val uintSbEquals= out(uint4 === uint8)
    val uintSbNotEquals = out(uint4 =/= uint8)

    val uintSbAdd = out(uint4 + uint8)
    val uintSbSub = out(uint4 - uint8)
    val uintSbMul = out(uint4 * uint8)
    val uintSbDiv = out(uint4 / uint8NotZero)
    val uintSbRem = out(uint4 % uint8NotZero)

    val uintSbAnd = out(uint4.resized & uint8)
    val uintSbOr = out(uint4.resized  | uint8)
    val uintSbXor = out(uint4.resized  ^ uint8)

    val uintSbSmaller = out(uint4 < uint8)
    val uintSbSmallerEquals = out(uint4 <= uint8)
    val uintSbBigger = out(uint4 > uint8)
    val uintSbBiggerEquals = out(uint4 >= uint8)

    val sintSbEquals= out(sint4 === sint8)
    val sintSbNotEquals = out(sint4 =/= sint8)

    val sintSbAdd = out(sint4 + sint8)
    val sintSbSub = out(sint4 - sint8)
    val sintSbMul = out(sint4 * sint8)
    val sintSbDiv = out(sint4 / sint8NotZero)
    val sintSbRem = out(sint4 % sint8NotZero)

    val sintSbAnd = out(sint4.resized & sint8)
    val sintSbOr = out(sint4.resized | sint8)
    val sintSbXor = out(sint4.resized ^ sint8)

    val sintSbSmaller = out(sint4 < sint8)
    val sintSbSmallerEquals = out(sint4 <= sint8)
    val sintSbBigger = out(sint4 > sint8)
    val sintSbBiggerEquals = out(sint4 >= sint8)



    val bitsSbEquals= out(bits4.resized === bits8)
    val bitsSbNotEquals = out(bits4.resized =/= bits8)

    val bitsSbAnd = out(bits4.resized & bits8)
    val bitsSbOr = out(bits4.resized | bits8)
    val bitsSbXor = out(bits4.resized ^ bits8)



    val uintBsEquals= out(uint8 === uint4)
    val uintBsNotEquals = out(uint8 =/= uint4)

    val uintBsAdd = out(uint8 + uint4)
    val uintBsSub = out(uint8 - uint4)
    val uintBsMul = out(uint8 * uint4)
    val uintBsDiv = out(uint8 / uint4NotZero)
    val uintBsRem = out(uint8 % uint4NotZero)

    val uintBsAnd = out(uint8 & uint4.resized)
    val uintBsOr = out(uint8 | uint4.resized)
    val uintBsXor = out(uint8 ^ uint4.resized)
    val uintBsNot = out(~uint8)

    val uintBsSmaller = out(uint8 < uint4)
    val uintBsSmallerEquals = out(uint8 <= uint4)
    val uintBsBigger = out(uint8 > uint4)
    val uintBsBiggerEquals = out(uint8 >= uint4)

    val sintBsEquals= out(sint8 === sint4)
    val sintBsNotEquals = out(sint8 =/= sint4)

    val sintBsAdd = out(sint8 + sint4)
    val sintBsSub = out(sint8 - sint4)
    val sintBsMul = out(sint8 * sint4)
    val sintBsDiv = out(sint8 / sint4NotZero)
    val sintBsRem = out(sint8 % sint4NotZero)

    val sintBsAnd = out(sint8 & sint4.resized)
    val sintBsOr = out(sint8 | sint4.resized)
    val sintBsXor = out(sint8 ^ sint4.resized)
    val sintBsNot = out(~sint8)

    val sintBsSmaller = out(sint8 < sint4)
    val sintBsSmallerEquals = out(sint8 <= sint4)
    val sintBsBigger = out(sint8 > sint4)
    val sintBsBiggerEquals = out(sint8 >= sint4)

    val bitsBsEquals= out(bits8 === bits4.resized)
    val bitsBsNotEquals = out(bits8 =/= bits4.resized)

    val bitsBsAnd = out(bits8 & bits4.resized)
    val bitsBsOr = out(bits8 | bits4.resized)
    val bitsBsXor = out(bits8 ^ bits4.resized)
    val bitsBsNot = out(~bits8)

    val bitsCat = out(bits8 ## bits4)





    val boolEquals    = out(boolA === boolB)
    val boolNotEquals = out(boolA =/= boolB)
    
    val boolAnd = out(boolA & boolB)
    val boolOr = out(boolA | boolB)
    val boolXor = out(boolA ^ boolB)
    val boolNot = out(!boolA)



    val uintAsBits = out(uint8.asBits)
    val uintAsSint = out(uint8.asSInt)

    val sintAsBits = out(sint8.asBits)
    val sintAsUint = out(sint8.asUInt)

    val bitsAsUint = out(bits8.asUInt)
    val bitsAsSint = out(bits8.asSInt)

    val boolAsBits = out(boolA.asBits)
    val boolAsUInt = out(boolA.asUInt)
    val boolAsSInt = out(boolA.asSInt)



    val bitsResizeBigger  = out(bits8.resize(16))
    val bitsResizeSmaller = out(bits8.resize(4))

    val uintResizeBigger  = out(uint8.resize(16))
    val uintResizeSmaller = out(uint8.resize(4))

    val sintResizeBigger  = out(sint8.resize(16))
    val sintResizeSmaller = out(sint8.resize(4))




    val stateNative,stateNativeMuxInternal = State(native)
    stateNative.assignFromBits(bits8(1 downto 0))
    val stateNativeBits   = out(stateNative.asBits)
    val stateNativeIsA    = out(stateNative === State.a)
    val stateNativeIsB    = out(stateNative === State.b)
    val stateNativeIsC    = out(stateNative === State.c)
    val stateNativeIsNotA = out(stateNative =/= State.a)
    val stateNativeIsNotB = out(stateNative =/= State.b)
    val stateNativeIsNotC = out(stateNative =/= State.c)
    stateNativeMuxInternal := Mux(boolC,State.a,State.b)
    val stateNativeMux = out(stateNativeMuxInternal.asBits)


    val stateBinarySequancial,stateBinarySequancialMux = State(binarySequential)
    stateBinarySequancial.assignFromBits(bits8(1 downto 0))
    val stateBinarySequancialBits   = out(stateBinarySequancial.asBits)
    val stateBinarySequancialIsA    = out(stateBinarySequancial === State.a)
    val stateBinarySequancialIsB    = out(stateBinarySequancial === State.b)
    val stateBinarySequancialIsC    = out(stateBinarySequancial === State.c)
    val stateBinarySequancialIsNotA = out(stateBinarySequancial =/= State.a)
    val stateBinarySequancialIsNotB = out(stateBinarySequancial =/= State.b)
    val stateBinarySequancialIsNotC = out(stateBinarySequancial =/= State.c)
    stateBinarySequancialMux    := Mux(boolC,State.a,State.b)
    stateBinarySequancialMux.asOutput()

    val stateBinaryOneHot,stateBinaryOneHotMux = State(binaryOneHot)
    stateBinaryOneHot.assignFromBits(bits8(2 downto 0))
    val stateBinaryOneHotBits = out(stateBinaryOneHot.asBits)
    val stateBinaryOneHotIsA    = out(stateBinaryOneHot === State.a)
    val stateBinaryOneHotIsB    = out(stateBinaryOneHot === State.b)
    val stateBinaryOneHotIsC    = out(stateBinaryOneHot === State.c)
    val stateBinaryOneHotIsNotA = out(stateBinaryOneHot =/= State.a)
    val stateBinaryOneHotIsNotB = out(stateBinaryOneHot =/= State.b)
    val stateBinaryOneHotIsNotC = out(stateBinaryOneHot =/= State.c)
    stateBinaryOneHotMux := (Mux(boolC,State.a,State.b))
    stateBinaryOneHotMux.asOutput()



    val bitsAggregateFixed = out(B(7 -> false,(6 downto 5) -> true,(4 downto 3) -> bits8(1 downto 0),(2 downto 1) -> boolA,0 -> True))
    val uintAggregateFixed = out(U(7 -> false,(6 downto 5) -> true,(4 downto 3) -> uint8(1 downto 0),(2 downto 1) -> boolA,0 -> True))
    val sintAggregateFixed = out(S(7 -> false,(6 downto 5) -> true,(4 downto 3) -> sint8(1 downto 0),(2 downto 1) -> boolA,0 -> True))

    val bitsAggregateUnfixedWidthFixedDefault = out(B(7 -> false,(6 downto 5) -> true,(4 downto 3) -> bits8(1 downto 0),0 -> True,(2 downto 1) -> true))
    val uintAggregateUnfixedWidthFixedDefault = out(U(7 -> false,(6 downto 5) -> true,(4 downto 3) -> uint8(1 downto 0),0 -> True,(2 downto 1) -> true))
    val sintAggregateUnfixedWidthFixedDefault = out(S(7 -> false,(6 downto 5) -> true,(4 downto 3) -> sint8(1 downto 0),0 -> True,(2 downto 1) -> true))

    val bitsAggregateUnfixedWidthUnfixedDefault = out Bits(8 bits)
    val uintAggregateUnfixedWidthUnfixedDefault = out UInt(8 bits)
    val sintAggregateUnfixedWidthUnfixedDefault = out SInt(8 bits)

    bitsAggregateUnfixedWidthUnfixedDefault := (7 -> false,(6 downto 5) -> true,(4 downto 3) -> bits8(1 downto 0),0 -> True,default -> boolA)
    uintAggregateUnfixedWidthUnfixedDefault := (7 -> false,(6 downto 5) -> true,(4 downto 3) -> uint8(1 downto 0),0 -> True,default -> boolA)
    sintAggregateUnfixedWidthUnfixedDefault := (7 -> false,(6 downto 5) -> true,(4 downto 3) -> sint8(1 downto 0),0 -> True,default -> boolA)




    val bitsRotateLeftUInt = out(bits32(26 downto 0).rotateLeft(uint8(4 downto 0)))
    val uintRotateLeftUInt = out(bits32(26 downto 0).asUInt.rotateLeft(uint8(4 downto 0)))
    val sintRotateLeftUInt = out(bits32(26 downto 0).asSInt.rotateLeft(uint8(4 downto 0)))
    val bitsRotateRightUInt = out(bits32(26 downto 0).rotateRight(uint8(4 downto 0)))
    val uintRotateRightUInt = out(bits32(26 downto 0).asUInt.rotateRight(uint8(4 downto 0)))
    val sintRotateRightUInt = out(bits32(26 downto 0).asSInt.rotateRight(uint8(4 downto 0)))
  }
}

class OperatorTesterCocotbBoot extends SpinalTesterCocotbBase {
  override def getName: String = "OperatorTester"
  override def createToplevel: Component =  new OperatorTester.OperatorTester
  override def pythonTestLocation: String = "tester/src/test/python/spinal/OperatorTester"
}

//object OperatorTesterGenVhdl{
//  def main(args: Array[String]) {
//    SpinalVhdl(new OperatorTester)
//  }
//}